File: src/estimator.js
const data = {
  region: {
    name: 'Africa',
    avgAge: 19.7,
    avgDailyIncomeInUSD: 4,
    avgDailyIncomePopulation: 0.73
  },
  periodType: 'days',
  timeToElapse: 38,
  reportedCases: 2747,
  population: 92931687,
  totalHospitalBeds: 678874
};

const convertToDays = (periodType, timeToElapse) => {
  switch (periodType) {
    case 'days':
      return timeToElapse;
    case 'weeks':
      return timeToElapse * 7;
    case 'months':
      return timeToElapse * 30;
    default:
      return timeToElapse;
  }
};

const fifteenPercent = (infectionsByRequestedTime) => 0.15 * infectionsByRequestedTime;

const availableBeds = (totalHospitalBeds, severeCasesByRequestedTime) => {
  const availableBedSpace = 0.35 * totalHospitalBeds;
  return Math.trunc(availableBedSpace - severeCasesByRequestedTime);
};

const ICUcare = (infectionsByRequestedTime) => Math.trunc(0.05 * infectionsByRequestedTime);

const ventilators = (infectionsByRequestedTime) => Math.trunc(0.02 * infectionsByRequestedTime);

const dollarsInFlightCalc = (
  infectionsByRequestedTime,
  avgDailyIncome,
  avgDailyIncomePercent,
  days
) => {
  const value = infectionsByRequestedTime * avgDailyIncome * avgDailyIncomePercent * days;
  return parseFloat(value.toFixed(2));
};

const covid19ImpactEstimator = (data) => {
  const outputData = {
    data: null,
    estimate: {
      impact: {},
      severeImpact: {}
    }
  };

  const {
    timeToElapse, reportedCases, periodType, totalHospitalBeds
  } = data;
  const { avgDailyIncomeInUSD, avgDailyIncomePopulation } = data.region;
  const { impact, severeImpact } = outputData.estimate;
  outputData.data = data;
  impact.currentlyInfected = reportedCases * 10;
  severeImpact.currentlyInfected = reportedCases * 50;
  const days = convertToDays(periodType, timeToElapse);
  const factor = 2 ** Math.trunc(days / 3);
  impact.infectionsByRequestedTime = impact.currentlyInfected * factor;
  severeImpact.infectionsByRequestedTime = severeImpact.currentlyInfected * factor;
  impact.severeCasesByRequestedTime = fifteenPercent(
    impact.infectionsByRequestedTime
  );
  severeImpact.severeCasesByRequestedTime = fifteenPercent(
    severeImpact.infectionsByRequestedTime
  );
  impact.hospitalBedsByRequestedTime = availableBeds(
    totalHospitalBeds,
    impact.severeCasesByRequestedTime
  );
  severeImpact.hospitalBedsByRequestedTime = availableBeds(
    totalHospitalBeds,
    severeImpact.severeCasesByRequestedTime
  );
  impact.casesForICUByRequestedTime = ICUcare(impact.infectionsByRequestedTime);
  severeImpact.casesForICUByRequestedTime = ICUcare(
    severeImpact.infectionsByRequestedTime
  );
  impact.casesForVentilatorsByRequestedTime = ventilators(
    impact.infectionsByRequestedTime
  );
  severeImpact.casesForVentilatorsByRequestedTime = ventilators(
    severeImpact.infectionsByRequestedTime
  );
  impact.dollarsInFlight = dollarsInFlightCalc(
    impact.infectionsByRequestedTime,
    avgDailyIncomePopulation,
    avgDailyIncomeInUSD,
    days
  );
  severeImpact.dollarsInFlight = dollarsInFlightCalc(
    severeImpact.infectionsByRequestedTime,
    avgDailyIncomePopulation,
    avgDailyIncomeInUSD,
    days
  );

  return outputData;
};

console.log(covid19ImpactEstimator(data));
// export default covid19ImpactEstimator;

File: on_covid19/estimator.py

def estimator(data):

    periodtype = data['periodType']
    time_to_elapse = data['timeToElapse']
    reported_cases = data['reportedCases']
    hospital_beds = data['totalHospitalBeds']
    income = data['region']['avgDailyIncomeInUSD']
    income_population = data['region']['avgDailyIncomePopulation']

    impact_cases = reported_cases * 10
    severe_cases = reported_cases * 50
    available_beds = 0.35 * hospital_beds

    # calculate the number of days

    def number_of_days(periodtype, time_to_elapse):
        if periodtype == 'days':
            days = time_to_elapse
        elif periodtype == 'weeks':
            days = time_to_elapse * 7
        elif periodtype == 'months':
            days = time_to_elapse * 30
        return days

    # infections by requested time, where case = currentlyInfected
    def infected_till_date(case):
        days = number_of_days(periodtype, time_to_elapse) // 3
        return case * (2 ** days)

    def severe_by_req_time(case):
        return 0.15 * infected_till_date(case)

    def hospital_bedsby_time(case):
        return int(available_beds - severe_by_req_time(case))

    def money_lost(case):
        inf = infected_till_date(case)
        days = number_of_days(periodtype, time_to_elapse)
        return int((inf * income_population * income) / days)
        # return round(inf * income_population * income * days, 2)

    result = {
        "data": data,
        "impact": {
            "currentlyInfected": impact_cases,
            "infectionsByRequestedTime": infected_till_date(impact_cases),
            "severeCasesByRequestedTime": int(severe_by_req_time(impact_cases)),
            "hospitalBedsByRequestedTime": hospital_bedsby_time(impact_cases),
            "casesForICUByRequestedTime": int(0.05 * infected_till_date(impact_cases)),
            "casesForVentilatorsByRequestedTime": int(0.02 * infected_till_date(impact_cases)),
            "dollarsInFlight": money_lost(impact_cases)
        },
        "severeImpact": {
            "currentlyInfected": severe_cases,
            "infectionsByRequestedTime": infected_till_date(severe_cases),
            "severeCasesByRequestedTime": int(severe_by_req_time(severe_cases)),
            "hospitalBedsByRequestedTime": hospital_bedsby_time(severe_cases),
            "casesForICUByRequestedTime": int(0.05 * infected_till_date(severe_cases)),
            "casesForVentilatorsByRequestedTime": int(0.02 * infected_till_date(severe_cases)),
            "dollarsInFlight": money_lost(severe_cases)
        }
    }
    return result
